ceedling 1.0.0 release candidate

mvandervoord opened the pull request:
I'm going to start building the 0.32 release candidate here. It might be a bit early for a PR, but this makes it easier to test on Actions.

Comments, help, and thoughts are welcome.

swaldhoer commented:
@mvandervoord will you actively ask for feedback before merging this/creating a release?

I would be interested to provide feedback once this pull request reaches a good stability from your point of view.

mvandervoord commented:
My purpose of labeling this branching so obviously was to solicit feedback. Let it roll in anytime :)

lukzeg commented on spec/gcov/gcov_test_cases_spec.rb:
```diff
@@ -194,4 +194,147 @@ def can_create_html_report
+
+        add_line = false
+        updated_prj_yml = []
+        File.read('project.yml').split("\n").each do |line|
```
Hi @mvandervoord ,

May I suggest to change this part responsible by updating project.yml to look:

```
File.read('project.yml').split("\n").each do |line|
          if line =~ /\:use_backtrace_gdb_reporter\:/
            line = line.gsub("FALSE", "TRUE")
          end
          updated_prj_yml.append(line)
        end
```

mvandervoord replied:
Thank your for the advice! I'm rolling this into a helper. :)

lukzeg commented on spec/gcov/gcov_test_cases_spec.rb:
```diff
@@ -194,4 +194,147 @@ def can_create_html_report
+
+        add_line = false
+        updated_prj_yml = []
+        File.read('project.yml').split("\n").each do |line|
```
Hi @mvandervoord ,

May I suggest to change this part responsible by updating project.yml to look:

```
File.read('project.yml').split("\n").each do |line|
          if line =~ /\:use_backtrace_gdb_reporter\:/
            line = line.gsub("FALSE", "TRUE")
          end
          updated_prj_yml.append(line)
        end
```

lukzeg commented on spec/spec_system_helper.rb:
```diff
@@ -874,11 +873,140 @@ def test_run_of_projects_fail_because_of_sigsegv_with_report
+
+        add_line = false
+        updated_prj_yml = []
+        File.read('project.yml').split("\n").each do |line|
```
Hi @mvandervoord ,

May I suggest to change this part responsible by updating project.yml to look:

```
File.read('project.yml').split("\n").each do |line|
          if line =~ /\:use_backtrace_gdb_reporter\:/
            line = line.gsub("FALSE", "TRUE")
          end
          updated_prj_yml.append(line)
        end
```

lukzeg commented on spec/spec_system_helper.rb:
```diff
@@ -874,11 +873,140 @@ def test_run_of_projects_fail_because_of_sigsegv_with_report
+
+        add_line = false
+        updated_prj_yml = []
+        File.read('project.yml').split("\n").each do |line|
```
Hi @mvandervoord ,

May I suggest to change this part responsible by updating project.yml to look:

```
File.read('project.yml').split("\n").each do |line|
          if line =~ /\:use_backtrace_gdb_reporter\:/
            line = line.gsub("FALSE", "TRUE")
          end
          updated_prj_yml.append(line)
        end
```

lukzeg commented on spec/spec_system_helper.rb:
```diff
@@ -874,11 +873,140 @@ def test_run_of_projects_fail_because_of_sigsegv_with_report
+
+        add_line = false
+        updated_prj_yml = []
+        File.read('project.yml').split("\n").each do |line|
```
Hi @mvandervoord ,

May I suggest to change this part responsible by updating project.yml to look:

```
File.read('project.yml').split("\n").each do |line|
          if line =~ /\:use_backtrace_gdb_reporter\:/
            line = line.gsub("FALSE", "TRUE")
          end
          updated_prj_yml.append(line)
        end
```

lukzeg reviewed (commented):
Hi @mvandervoord ,

I investigated why tests are failing on CI and mostly looks like the changes applied in project.yml used under test break them. I added small fix in my comments, have hope that it will help you.

deltalejo commented:
Hi!
This version will include the [improved support for separate out directories](https://github.com/ThrowTheSwitch/Ceedling/pull/591)?

mvandervoord commented:
@swaldhoer -- thanks for doing some testing!

First, for future reference, when posting a block of code on a github discussion like this, you can put the block of code (or command prompt response, in this case) between lines that contain nothing but three backticks like "```". This will make a block that preserves all the whitespace, unlike the single backticks you were using above. :)

```
# ```
# like this, without the comments at the beginning
# ```
```

On to the real topics!

First, do I understand correctly that you originally created your project to install to the vendor folder? How did you install the latest test release? Did you just download and copy over the vendor directory? This is going to make a weird little mix where you're doing your command parsing with the old version of ceedling, but using the new version in the guts. It should mostly work, but might create some quirks.

This problem is my fault. I should create a gem for you to be able to test with. I'll do so soon and post it for people to try. Then, once the new gem is installed, you can do `ceedling upgrade <dirname>` from just outside the project folder to upgrade it.

Finally, once we have you running a full new version, if you run into problems, if you could use `ceedling verbosity[5] clobber test:all` for when you capture problems, it'd be good to see all the details.

Thanks again! I should have a gem posted soon.

Letme commented:
@mvandervoord can you put the gem creation in the CI, so that we can simply download it here? Should be much easier and requires you not to push it anywhere.

mvandervoord commented:
@Letme -- that's a great idea. I'll see if Github Actions has that option :)

swaldhoer commented:
> First, do I understand correctly that you originally created your project to install to the vendor folder? How did you install the latest test release? Did you just download and copy over the vendor directory? This is going to make a weird little mix where you're doing your command parsing with the old version of ceedling, but using the new version in the guts. It should mostly work, but might create some quirks.

I installed the required gems from the commit c98953e89c0047b6716565a6dcb1a7f354d1df6a. And then updated our ceedling copy.
We wrap ceedling into our build tool, so that there is only one tool to run everything from an top level view.
Basically our setup is like this:
- `./build/unit_test`: this directory is created by the build tool, from there ceedling is run and is where the `project.yml` is located
- `tools/vendor/ceedling/*`: our ceedling copy

So we basically use the vendor approach, but some paths are tweaked to match our project setup. If you are interested I can further explain.

> This problem is my fault. I should create a gem for you to be able to test with. I'll do so soon and post it for people to try. Then, once the new gem is installed, you can do `ceedling upgrade <dirname>` from just outside the project folder to upgrade it.
>
> Finally, once we have you running a full new version, if you run into problems, if you could use `ceedling verbosity[5] clobber test:all` for when you capture problems, it'd be good to see all the details.
>
> Thanks again! I should have a gem posted soon.

Ok, Then I will wait for a new gem and test again. I could overcome the include issue by adding `../../tools/vendor/ceedling/vendor/cmock/src` etc. to the include paths. But with 0.31.1, there was no reason to add these include paths, and ceedling somehow magical knew where to search for the include paths.
However, the build than ran for ever and I needed to cancel it.

I will only be able to run the next next thursday (2023-02-16), but I will do for sure and post an update.
We are very willing to keep our dependencies up to date and if this helps to make this release better, we are glad to help :)

mvandervoord commented:
@swaldhoer -- There's a prerelease gem posted in releases: https://github.com/ThrowTheSwitch/Ceedling/releases/download/0.32.0-665b570/ceedling-0.32.0-665b570.gem

@Letme -- Thanks for the excellent idea. I'm liking automatic prereleases.

swaldhoer commented:
> @swaldhoer -- There's a prerelease gem posted in releases: https://github.com/ThrowTheSwitch/Ceedling/releases/download/0.32.0-665b570/ceedling-0.32.0-665b570.gem
> 
> @Letme -- Thanks for the excellent idea. I'm liking automatic prereleases.

Sorry this is just a memory dump from friday. I used the gem you linked and then updated the vendored files using the cli arguments.
- Starting from a clean repo state. (``git clean -xdf``) and then running ceedling, errors out, as ceedling calls the compiler and expects some file ``foo/_temp/abc.c`` and the file is not yet created.
- re-running the command again, ceedling creates 100% cpu usage, takes quite some time, and then misses some headers (the same mentioned above). By adding the ceedling vendored headers to the list of includes in the ``project.yml`` (e.g., ``../../tools/vendor/ceedling/vendor/cmock/src``), the headers are found.
- re-running ceedling again takes for ever (longer than with the current release) and then fails to compile the first(!) test case.

lukzeg commented:
Hi @swaldhoer ,

Do you know maybe which header file is causing mentioned by you problem? As mostly I had similar issue in the past similar to you, when one header file was causing for me similar behavior.

Finding this file and sharing project.yml will help to investigate what change cause this issue.

swaldhoer commented:
> Hi @swaldhoer ,
> 
> Do you know maybe which header file is causing mentioned by you problem? As mostly I had similar issue in the past as you. It was connected with multiple of the pre-processor statements.
> 
> Finding this file and sharing project.yml will help to investigate what change cause this issue.
> 
> 

Unfortunately the answer is simple: all header files that are vendores by ceedling.

I will share the project yaml later/tomorrow.

swaldhoer commented:
Here is the project configuration. Imporant:
- the project files is in `path/to/repo/build/unit_test`
- the vendored ceedling is in `path/to/repo/tools/vendor/ceedling`. To avoid any ambiguity about the vendoring: the ceedling binary is then at `path/to/repo/tools/vendor/ceedling/bin/ceedling`

<details>
<summary>project.yml</summary>

```yaml
:project:
  :use_exceptions: TRUE
  :use_test_preprocessor: TRUE
  :use_mocks: TRUE
  :use_auxiliary_dependencies: TRUE
  :build_root: .
  :test_file_prefix: test_
  :which_ceedling: ../../tools/vendor/ceedling
  :ceedling_version: 0.31.1
  :default_tasks:
    - test:all

:environment: []

:extension:
  :executable: .out

:paths:
  :test:
    - +:../../tests/unit/**
    - -:../../tests/unit/support
  :source:
    - ../../src/app/**
    - ../../src/opt/**
  :include:
    - +:./include/**
    - +:../../src/**
    - -:../../src/app/**
    - -:../../src/opt/**
  :support:
    - ../../tests/unit/support/**

:defines:
  :common: &common_defines
    - UNITY_INCLUDE_EXEC_TIME
  :test:
    - *common_defines
  :test_preprocess:
    - *common_defines
    - INC_FREERTOS_H

:cmock:
  :when_no_prototypes: :warn
  :enforce_strict_ordering: TRUE
  :mock_prefix: Mock
  :weak: ""
  :strippables:
    [
      "(.FREERTOS_SYSTEM_CALL)",
      "(.PRIVILEGED_FUNCTION)",
      "(portDONT_DISCARD)",
    ]
  :includes:
    - "FreeRTOSConfig.h"
    - "FreeRTOS.h"
    - "portmacro.h"
    - "mpu_wrappers.h"
    - "portable.h"
    - "task.h"
    - "queue.h"
    - "semphr.h"
    - "stream_buffer.h"
    - "event_groups.h"
    - "string.h"
  :plugins:
    - :ignore
    - :callback
    - :ignore_arg
    - :return_thru_ptr
  :treat_externs: :include
  :treat_as:
    uint8: HEX8
    uint16: HEX16
    uint32: UINT32
    int8: INT8
    bool: UINT8

:tools_test_linker:
  :arguments:
    - -lm
    - -flto
:tools_gcov_linker:
  :arguments:
    - -lm
    - -flto
:gcov:
  :utilities:
    - gcovr
  :reports:
    - HtmlDetailed
    - Text
    - Cobertura
  :abort_on_uncovered: true
  :gcovr:
    :report_root: "../../"
    :report_exclude: ".*vendor.*|.*build.*|.*test.*|.*tests.*|.*lib.*|.*Test.*|.*src.hal.*|.*src.os.*"
    :exclude_directories: ".*tests.*|.*src.os.*|.*build.*"
    :sort_percentage: true
    :sort_uncovered: false
    :html_medium_threshold: 60
    :html_high_threshold: 85
    :print_summary: true
    :num_parallel_threads: 4
    :keep: false

:junit_tests_report:
  :artifact_filename: report_junit.xml

:plugins:
  :load_paths:
    - ../../tools/vendor/ceedling/plugins
  :enabled:
    - gcov
    - stdout_pretty_tests_report
    - module_generator
    - xml_tests_report
    - junit_tests_report
```

</details>

mvandervoord commented:
@swaldhoer -- it sounds like the new automatic generation of the gem isn't adding the submodules (unity, cmock, etc) to the gem. I'll figure out what's going on with that.

More concerning are the other failures. It's looking like something with the updated preprocessor calls isn't happy? Can you do a quick test where you disable the preprocessor and `rake clobber test:all` to see if it will complete a test then?

As a side note, feel free to add the following two lines to your project section:

```
:project:
  :test_threads: 8
  :compile_threads: 8
```

This should help speed things up. :)

swaldhoer commented:
I did not use the update mechanism. I created a new project and then copied everything. This is how we did it (1 or 2 times I guess) and it worked.
Doing so again, I saw changes in the vendored dependencies at ``tools/vendor/ceedling/vendor/cmock`` etc., so I assume this was fine.

mvandervoord commented:
Hahaha... you are correct. I meant `ceedling clobber test:all`, not `rake`. 

Your upgrade method should be fine. That suggests I have something wrong in the way ceedling is finding its subprojects when not using the gem. Thanks for the help! I should be able to find this error. :)

Letme commented:
I am updating to newer version (because github default dockers now use Ruby 3.2) and just hit the following problem when using this branch.

I have following `rakefile` while Ceedling is submoduled to `tools/ceedling`:

```
PROJECT_CEEDLING_ROOT = "tools/ceedling"
load "#{PROJECT_CEEDLING_ROOT}/lib/ceedling/rakefile.rb"

task :default => %w[ test:all release ]
```

And the error I get  after running command: `rake -m -j 4 options:gcc clobber test:all` (I have `gcc.yml` in `targets/`)
```
rake aborted!
NameError: uninitialized constant Configurator::Ceedling
*/tools/ceedling/lib/ceedling/configurator.rb:178:in `find_and_merge_plugins'
*/tools/ceedling/lib/ceedling/setupinator.rb:26:in `do_setup'
tools/ceedling/lib/ceedling/rakefile.rb:46:in `<top (required)>'
*/rakefile:5:in `load'
*/rakefile:5:in `<top (required)>'
```

Any ideas what is wrong?

deltalejo commented:
Have you tried what is stated [here](https://github.com/ThrowTheSwitch/Ceedling/blob/master/docs/CeedlingPacket.md#adding-handy-rake-tasks-for-your-project-without-fancy-pants-custom-plugins)?
It seems that the `rakefile` should be something like

```ruby
PROJECT_CEEDLING_ROOT = "tools/ceedling"
load "#{PROJECT_CEEDLING_ROOT}/lib/ceedling.rb"
Ceedling.load_project

task :default => %w[ test:all release ]
```

Letme commented:
> @Letme are you calling `rake` instead of `ceedling` ?

Yes, I am. Because I do not install ceedling, i use it as submodule. I guess i need to call ./tools/ceedling/bin/ceedling then... (or something- sorry am on mobile)

Letme commented:
OK, if I change `rakefile` with what @deltalejo suggested, it takes the local ceedling in submodule (but I also installed the above gem) I get it to build, but then building fails:

```
Building Objects
----------------
Compiling TestDSP_runner.c...
#<Thread:0x00007f4f4eb0f528 /*/tmp/mlx90632-library/tools/ceedling/lib/ceedling/par_map.rb:7 run> terminated with exception (report_on_exception is true):
/*/mlx90632-library/tools/ceedling/lib/ceedling/flaginator.rb:23:in `to_h': wrong element type String at 0 (expected array) (TypeError)

  hash.partition(&predicate).map(&:to_h)
                                 ^^^^^^
	from /*/mlx90632-library/tools/ceedling/lib/ceedling/flaginator.rb:23:in `map'
	from /*/mlx90632-library/tools/ceedling/lib/ceedling/flaginator.rb:23:in `partition'
	from /*/mlx90632-library/tools/ceedling/lib/ceedling/flaginator.rb:34:in `get_flag'
	from /*/mlx90632-library/tools/ceedling/lib/ceedling/flaginator.rb:71:in `flag_down'
	from /*/mlx90632-library/tools/ceedling/lib/ceedling/generator.rb:104:in `generate_object_file'
	from /*/mlx90632-library/tools/ceedling/lib/ceedling/test_invoker_helper.rb:61:in `block in generate_objects_now'
	from /*/mlx90632-library/tools/ceedling/lib/ceedling/par_map.rb:10:in `block (2 levels) in par_map'
rake aborted!
TypeError: wrong element type String at 0 (expected array)
/*/mlx90632-library/tools/ceedling/lib/ceedling/flaginator.rb:23:in `to_h'
/*mlx90632-library/tools/ceedling/lib/ceedling/flaginator.rb:23:in `map'
/*/mlx90632-library/tools/ceedling/lib/ceedling/flaginator.rb:23:in `partition'
/*/mlx90632-library/tools/ceedling/lib/ceedling/flaginator.rb:34:in `get_flag'
/*/mlx90632-library/tools/ceedling/lib/ceedling/flaginator.rb:71:in `flag_down'
/*/mlx90632-library/tools/ceedling/lib/ceedling/generator.rb:104:in `generate_object_file'
/*/mlx90632-library/tools/ceedling/lib/ceedling/test_invoker_helper.rb:61:in `block in generate_objects_now'
/*/mlx90632-library/tools/ceedling/lib/ceedling/par_map.rb:10:in `block (2 levels) in par_map'
Tasks: TOP => test:all
```

I also ran `ceedling upgrade mlx90632-library`, but also the ceedling in vendor repo fails.

File is: https://github.com/melexis/mlx90632-library/blob/master/test/TestDSP.c

Letme commented:
OK, found the change. Flags definition was adjusted meanwhile so:

```
 :flags:
   :test:
     :compile:
      - -Wall
```
needs to become

```
 :flags:
   :test:
     :compile:
      :*:
        - -Wall
```

Letme commented:
Ok, I think there is another dependency. First I had to install gem in docker to get even above modification within `rakefile` running (some strange include problems). Secondly the new feature from #740 now requires to have gdb installed as otherwise it outputs: `ERROR: Config filepath [:tools][:backtrace_settings][:executable]['gdb'] does not exist in system search paths.` even though I have put `:use_backtrace_gdb_reporter: FALSE` in my `project.yml`. Locally I have gdb installed and this all works.

Letme commented:
Found a solution in main `project.yml` where I simply defined `backtrace_settings: optional: TRUE`. So now I need to figure if the command line argument `options:filename` still work.

mvandervoord commented:
@Letme I suspect you want `project:filename` not `options:filename`. The difference is so subtle I'm considering dropping the second option completely. The former does a merge of the yaml files, allowing you to override *most* settings you might be interested in overriding. The options version concentrates on just things like flags and defines and doesn't allow main settings or tools to change.

Letme commented:
Hm, for me `project` does not work in the expected way (doesn't merge).

I have a common configuration in project.yml and then I have tools specific and flags (merging stuff) in `targets/compilername.yml`. I think `project` expects whole configuration to be in one yml, because I also tried `ceedling project:project project:gcc clobber test:all` and it is not working in merging. So for me `options` merges configuration (i also do not need `-c {1}` `-o {2}` in that case in `tools:toolname:` section, while project is standalone.

Anyway, that is not discussion that should block the release of 0.32 . Can we get this merged and released soon (3 people already confirmed functionality, which is I think more than for previous releases).

swaldhoer commented:
Regarding the release: It does currently not work for us.
Idk if you already answered my mail @mvandervoord as I am on vacation with no access to my mails.

Until next week 🏖

mvandervoord commented:
@Letme -- There are a number of issues above that seem worth sorting out first. I hate to make a release where something major (like enabling preprocessing) breaks things for people. I'm unavailable for a couple weeks, but then plan to push on this release again.

Letme commented:
As mentioned in #758 there is an remaining issue with the `ceedling new <dirname>` command, which returns `EISDIR` error on Windows only. I have tried it on Linux and it works - so there must be something cross-platform.

swaldhoer commented on lib/ceedling/configurator_builder.rb:
```diff
@@ -13,19 +13,33 @@ class ConfiguratorBuilder
+      # Some key names can be C file names that can include dashes; dashes are not allowed in Ruby method names
+      # Downcase the key names and replace any illegal dashes with legal underscores
+      # Downcased key names create consistency and ensure no method names become Ruby constants by accident
+      eval("def #{key.to_s.gsub('-','_').downcase}() return @project_config_hash[:#{key}] end", context)
```
Minor comment: I have zero glue about Ruby, but for such "create a good name from arbitrary infput"-tasks in Python I always go by
````python
"".join([ c if c.isalnum() else "_" for c in s ]
```
so that I do not only need to implement every case. Maybe something like that is also a good solution here?

swaldhoer replied:
And for the special case of creating a function name, we would need to make sure that the first character is a letter, so, me might need to add a prefix.

tbates-redarc commented:
I would like to get #818 included in this release but as this branch is not currently passing CI, I have submitted the PR against `master`. Please let me know if/when you would like me to retarget it to this branch.

swaldhoer commented on plugins/gcov/README.md:
```diff
@@ -227,224 +543,288 @@ Generation of JSON reports may be modified with the following configuration item
+
+Details of interest for this plugin to be modified or made use of using 
+Ceedling's advanced features are primarily contained in 
+[defaults_gcov.rb](conig/defaults_gcov.rb) and [defaults.yml](config/defaults.yml).
```
```suggestion
[defaults_gcov.rb](config/defaults_gcov.rb) and [defaults.yml](config/defaults.yml).
```

swaldhoer commented on plugins/gcov/README.md:
```diff
@@ -227,224 +543,288 @@ Generation of JSON reports may be modified with the following configuration item
+Ceedling's advanced features are primarily contained in 
+[defaults_gcov.rb](conig/defaults_gcov.rb) and [defaults.yml](config/defaults.yml).
+
+## “gcvor not found”
```
```suggestion
## “gcovr not found”
```

swaldhoer commented on plugins/gcov/README.md:
```diff
@@ -227,224 +543,288 @@ Generation of JSON reports may be modified with the following configuration item
+## “gcvor not found”
+
+`gcovr` is a Python-based application. Depending on the particulars of its 
+installation and your platform, you may encounter a “gcvor not found” error. 
```
```suggestion
installation and your platform, you may encounter a “gcovr not found” error. 
```

swaldhoer commented on license.txt:
```diff
@@ -1,31 +1,24 @@
- copies of the Software, and to permit persons to whom the
- Software is furnished to do so, subject to the following
- conditions:
+https://opensource.org/license/mit/
```
Is this link here actually needed? Because if not, the license would be `MIT`, otherwise, i.e., including this link, the license is custom.

swaldhoer replied:
Fixed in f57f423bb7aaf870343be04413a68718518a4593.

erijo commented:
Trying version 0.32.0-352ef30 with ruby 3.2.3 (from Ubuntu 24.04) I get this warning:

```
/.../vendor/ceedling/plugins/fff/lib/fff.rb:90: warning: already initialized constant CMock
/.../vendor/ceedling/vendor/cmock/lib/cmock.rb:16: warning: previous definition of CMock was here
```

Adding `Object.send(:remove_const, :CMock)` in ffb.rb just before `CMock = FffCMockWrapper`  gets rid of the warning, but I'm not sure if that's the correct way to fix it.

lynch-andrew-m commented:
:pineapple: @mvandervoord, @mkarlesky, et al., what is the criteria for when a release is ready? Are there specific open issues you're looking to close on beforehand? Anything that community members could take a look at to help prepare the release?

I apologize in advance if this already exists somewhere... I couldn't find it

mvandervoord commented:
Thanks for asking, @lynch-andrew-m :)

We do have a small list of issues that we're working through before this release. Making it public is something we need to get better at. It includes the following: 

 - Issue #868 - making inlines work mocked headers 
 - PR #818 - Restructuring config load order
 - PR #856 - Improve segfault / backtrace handling
 
 - PR #783 - CMock sometimes selects wrong file (this is targeting the wrong branch -- master -- and therefore it's really a matter of verifying we've already fixed this elsewhere and closing it, or cherry-picking the important bits)

For all these, there are wonderful contributions from the community, but the contributions require some tweaking because they're aimed at a previous version. Some cherry-picking and refinement are needed because of things we've refactored in the process.

We would love to get the following PR's merged also (as above, these may need a bit of tweaking. Likely they need some testing and more documenting. They could potentially be good places for someone to jump in!) :

- PR #807 Valgrind plugin 
- PR #764 CppCheck plugin
- PR #776 Fix tools arrays (needs conflicts resolved and evaluation)

Letme commented:
@mvandervoord can you add labels to all those PRs/Issues, so that we have an up-to-date overview when stuff gets merged?

mvandervoord commented:
good idea!

M-Bab commented:
Thanks @mvandervoord for your great unit testing framework. I am very glad that the issue https://github.com/ThrowTheSwitch/Ceedling/issues/868 is getting picked up again for 0.32.

As a little reminder we had a lot of background info to this evaluated in https://github.com/ThrowTheSwitch/Ceedling/issues/706. I also prepared 2 MRs resolving the issue (https://github.com/ThrowTheSwitch/Ceedling/pull/728, https://github.com/ThrowTheSwitch/CMock/pull/421). This solution wouldn't quite win a beauty competition but we use it on several projects for over a year in my company now.

If I can be of any assistance getting this issue fixed let me know. I would be very glad if we can get rid of my custom solution once vanilla Ceedling 0.32 supports use_test_processor + treat_inlines.

Letme commented:
I want to see debug information (like command line what is passed to compiler/linker/etc.) and 0.x version that was
`ceedling verbosity[4] test:all`. This doesn't work in 1.0.0 pre-release. Also `ceedling logging` doesn't work anymore. Do we have a replacement for these debug commands?

Letme commented:
Found it. In `project.yml` under `:project:` directive you need to put `:verbosity: 4`. Might be a good idea to add that default value to the config (same as there is default value under cmock section).

mvandervoord commented:
@Letme also it's been move to a flag:

```
ceedling --verbosity=debug test:all
```

or

```
ceedlig -v=4 test:all
```

Letme commented:
Thanks for quick answer. Here is another finding, but I think this was raised in some issue (could not find it). When you have report option `report_tests_raw_output_log` plugin enabled, `gcov` task fails with:

```
OK
Plugin | Gcov > :post_test_fixture_execute...
Plugin | ReportTestsRawOutputLog > :post_test_fixture_execute...
Plugin | Gcov > :post_build...
🪲 ERROR: Exception raised in plugin `gcov` within build hook :post_build

Ceedling operations completed in 670 milliseconds

🧨 EXCEPTION: No test results report template has been set.
```

When I switch back to default `report_tests_pretty_stdout` it completes just fine.

Letme commented:
And here is anotherone. Lets assume you build your tests without coverage (so without gcov, gcovr), but your `project.yml` has gcov plugin enabled, your `ceedling test:all` command (which does not use gcov) will fail. I assume we should be loading plugins only when they are invoked, not that we load all plugins at the beginning and then even if they are not used we fail.

```
ceedling test:all
🪲 ERROR: Exception raised while trying to load plugin: gcov
🧨 EXCEPTION: default_gcov_gcovr_report ↳ :executable => `gcovr` does not exist in system search paths
🚧 Loaded project configuration at default location using ./project.yml
```

mvandervoord commented:
> Thanks for quick answer. Here is another finding, but I think this was raised in some issue (could not find it). When you have report option `report_tests_raw_output_log` plugin enabled, `gcov` task fails with:
> 
> ```
> OK
> Plugin | Gcov > :post_test_fixture_execute...
> Plugin | ReportTestsRawOutputLog > :post_test_fixture_execute...
> Plugin | Gcov > :post_build...
> 🪲 ERROR: Exception raised in plugin `gcov` within build hook :post_build
> 
> Ceedling operations completed in 670 milliseconds
> 
> 🧨 EXCEPTION: No test results report template has been set.
> ```
> 
> When I switch back to default `report_tests_pretty_stdout` it completes just fine.

Thanks. We'll look into this one! I really appreciate you testing the RC out! Thanks!

mvandervoord commented:
> And here is anotherone. Lets assume you build your tests without coverage (so without gcov, gcovr), but your `project.yml` has gcov plugin enabled, your `ceedling test:all` command (which does not use gcov) will fail. I assume we should be loading plugins only when they are invoked, not that we load all plugins at the beginning and then even if they are not used we fail.
> 
> ```
> ceedling test:all
> 🪲 ERROR: Exception raised while trying to load plugin: gcov
> 🧨 EXCEPTION: default_gcov_gcovr_report ↳ :executable => `gcovr` does not exist in system search paths
> 🚧 Loaded project configuration at default location using ./project.yml
> ```

We've changed the behavior of the gcov plugin a bit here. If you've enabled the gcov plugin in your project.yml file, it *always* builds tests with coverage. The separate `gcov:all` and `test:all` calls were confusing for everyone starting to use it, as was the need to rebuild everything when you switched between a coverage and non-coverage test call. So we've streamlined it all. If your project file says you want coverage, you get coverage... no extra commands or procedure changes required.

Does this make sense? 

Do you agree with this logic? Does it leave you with some concerns we may not have considered?

Letme commented:
> 
> Does this make sense?
> 
> Do you agree with this logic? Does it leave you with some concerns we may not have considered?

Based on your statements I have understood that now `test` and `gcov` are synonyms when `gcov` plugin is enabled. I have tested this by running `test:all` and it does not produce coverage reports when `gcov` is enabled in `project.yml`, which means output is different than `gcov:all`, so they are not completely the same. After re-reading I thought what you meant is that when you run `test` task you already create `gcda` `gcdo` files, but looking at `build` folder that is also not true (also they are separated, so I am not sure you will not need to rebuild, because `build/gcov/out/` and `build/test/out/` are different). So I am missing something.

My concern is that the additional flags passed to the compiler, to create coverage information, might cause a runtime failure in some cases due to more debug information added, so when they are split as they are now, they will make it easier to debug that its coverage information issue or real test issue. But it is also true I have compiler trust issues from past.

One more thing I did last night (to avoid this checking of gcov in path): I tried to use mixins based on documentation and failed. 

Here is how I set it up based on documentation:

`project.yml`:
```yaml
:mixins:
  :enabled: []
  :load_paths:
    - test/support

:plugins:
  :load_paths: []
  :enabled:
    #- beep                           # beeps when finished, so you don't waste time waiting for ceedling
    - module_generator                # handy for quickly creating source, header, and test templates
    #- gcov                            # test coverage using gcov. Requires gcc, gcov, and a coverage analyzer like gcovr
```

`test/support/coverage.yml` (tried also `test/support/coverage.yaml` and `coverage.yml`)
```yaml
:plugins:
  :load_paths: []
  :enabled:
    - gcov                            # test coverage using gcov. Requires gcc, gcov, and a coverage analyzer like gcovr

```

And commands I ran `ceedling --project=project.yml --mixin=coverage test:all`, `ceedling --project=project.yml --mixin=coverage.yml test:all`, `ceedling --project=project.yml --mixin=test/support/coverage.yml test:all` all return below error, which to me indicated that something in `mixin yml` is wrong. I remember there was an issue floating around about the order of how these were taken, but here it should not really matter because I did not provide multiple mixins, just one. And toggling that gcov line in `:plugins:` is basically all I need :wink: 

```
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.

🧨 EXCEPTION: undefined method `each' for an instance of String
Backtrace ==>
/home/cmo/.gem/gems/ceedling-1.0.0/bin/cli.rb:309:in `build': undefined method `each' for an instance of String (NoMethodError)
	/home/cmo/.gem/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	/home/cmo/.gem/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	/home/cmo/.gem/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	/home/cmo/.gem/gems/ceedling-1.0.0/bin/cli.rb:92:in `start'
	/home/cmo/.gem/gems/ceedling-1.0.0/bin/ceedling:125:in `<top (required)>'
	/home/cmo/.gem/bin/ceedling:25:in `load'
	/home/cmo/.gem/bin/ceedling:25:in `<main>'
```

mvandervoord commented:
my apologies. I had thought that we had decided to merge `gcov:all` and `test:all`, etc... it sounds like that didn't happen. In any case, there appear to be some issues remaining! Thanks for the sleuthing!

mkarlesky commented:
> Plugin | Gcov > :post_test_fixture_execute...
> Plugin | ReportTestsRawOutputLog > :post_test_fixture_execute...
> Plugin | Gcov > :post_build...
> 🪲 ERROR: Exception raised in plugin `gcov` within build hook :post_build
> 
> Ceedling operations completed in 670 milliseconds
> 
> 🧨 EXCEPTION: No test results report template has been set.
> ```
> 
> When I switch back to default `report_tests_pretty_stdout` it completes just fine.

@Letme This has been fixed in the latest prerelease build. You happened to be exercising a little used combination of no plugin-based test results report (it had nothing to do with the raw results log file report) with the GCov plugin. This problem has been lurking for a very long time!

Letme commented:
Cool. I am usually more "raw" guy, so maybe I bypassed that with deleting gcov default configuration (not sure what you did to fix it). Anyway, I can confirm 331a769 solved the problem.

mkarlesky commented:
> And here is anotherone. Lets assume you build your tests without coverage (so without gcov, gcovr), but your `project.yml` has gcov plugin enabled, your `ceedling test:all` command (which does not use gcov) will fail. I assume we should be loading plugins only when they are invoked, not that we load all plugins at the beginning and then even if they are not used we fail.
> 
> ```
> ceedling test:all
> 🪲 ERROR: Exception raised while trying to load plugin: gcov
> 🧨 EXCEPTION: default_gcov_gcovr_report ↳ :executable => `gcovr` does not exist in system search paths
> 🚧 Loaded project configuration at default location using ./project.yml
> ```

@Letme You are absolutely correct. The newest version of Ceedling validates tools within plugins conditionally — based on configuration. It so happens that the GCov plugin is a bit of a special case with its reliance on `gcov:` tasks. Generally, the assumption is that if a user enables a plugin, they intend to use it, and it's best to fail clearly and early if there's a tool problem. The GCov plugin abides by that rule. In fact, it does not actually validate the reporting tools unless reports are specified in the plugin's configuration. But, your point is taken. There's a bit of a gray zone on intent with actually executing coverage testing via a named task. So, I just added a check for `gcov:` tasks when the plugin is loaded that prevents tool validation if no `gcov:` task is requested. This is available in the latest prerelease.

(Inicidentally, we won't be able to do something better than the `gcov:` task convention until we fully excise Ceedling's reliable on Rake and totally replace the plugin architecture at that time.)

swaldhoer commented on plugins/gcov/lib/gcovr_reportinator.rb:
```diff
@@ -49,85 +91,50 @@ def make_reports(opts)
+  def args_builder_common(gcovr_opts)
     args = ""
-    args += "--root \"#{gcovr_opts[:report_root] || '.'}\" "
+    args += "--root \"#{gcovr_opts[:report_root]}\" " unless gcovr_opts[:report_root].nil?
```
Question: Might it make sense to remove all that custom parsing (from config variable names to `gcovr` argument names) into just reading an argument array from that config file that is passed to `gcovr` as arguments? This would make this plugin just work with any version of `gcovr`, reduce the code of this plugin and reduce the maintance work.

So instead of

```yaml
:gcov:
  :utilities:
    - gcovr
  :reports:
    - HtmlDetailed
  :gcovr:
    :report_root: "foo"
    :config_file: "bar"
```

we would just write:
```yaml
:gcov:
  :utilities:
    - gcovr
  :reports:
    - HtmlDetailed
  :gcovr: ["--root=foo", "--config=bar"]
```

mkarlesky replied:
@swaldhoer Maybe! The original plugin was expanded by community members to make use of `gcovr` _and_ `reportgenerator`. A good deal of the code in the plugin is trying to create a configuration in common between the two tools (whose functionality does overlap). These contributions did bring a big jump in functionality and seem rather popular. Maintenance and improvements have been a bear, though. After I don't know how many revisions over the last year, the GCovr plugin is finally in a semi-decent place… for more rounds of revisions! Perhaps a generous community member will try to spike out your idea to see how well it works. I know we won't have the bandwidth to try it before finally pushing out 1.0.0.

jhcasco commented:
I have a minor request regarding the documentation of TEST_SOURCE_FILE().  In the past, there have been ad-hoc solutions for testing static functions (the most common being re-defining the static keyword).  Our company's solution has been using the TEST_FILE() directive to test the static functions directly.  This is especially useful for static functions that work by modifying static global variables (e.g. state machines).  Attempting this via static keyword re-defines is a nightmare mix of STATIC_LOCAL, STATIC_GLOBAL, etc. to achieve the same functionality during test and really introduces a high amount of risk against our industry standards (MISRA C).

Could a use case for TEST_SOURCE_FILE() be added in the Ceedling_Packet.md to demonstrate this functionality officially?  Or is this a bigger philosophical question about design approach, unit definition, and testing side effects?

swaldhoer commented:
I am on vacation, but I can add how I use TEST_SOURCE_FILE  when I am back.

For testing static functions, I just add an *externalize* wrapper for the function under test that has the exact same declariation (except that it is extern of course) and just calls the static function. The function declaration and definition arw both hidden in the actual build behind the unit test marco. Works fine for a larger project.

mkarlesky commented:
@jhcasco @swaldhoer It hasn't happened yet, but we'd like to start a _Cookbook_ to document comment configuration + code solutions. This topic sounds like a good addition when it's time.

Would you both be willing to create Github gists to demonstrate what you're talking about? It'd be a lot easier to understand to see some sample source and test code along with any Ceedling configuration details.

jhcasco commented:
https://gist.github.com/jhcasco/73a131753261557b6c9e31a10414fd5c

This demonstrates current setup in Ceedling V0.31.1, though I seem to recall not running into *.gcno/*.gcna generation errors and more running into problems with gcovr requiring TEST_FILE() directives...  I guess this turned more into a cookbook for including a *.c directly.

Is something more official supported with the 1.0.0 release candidate?

swaldhoer commented:
Detailed version information seems not to work for local/vendored ceedling for (https://github.com/ThrowTheSwitch/Ceedling/pull/739/commits/ba45d2c0d2ba7da67ae09eb3b1d5e32f2561ded1):

```verbatim
Welcome to Ceedling!

   Ceedling => 1.0.0
      CMock => 2.5.4
      Unity => 2.6.0
CException => 1.3.4
```

I'll see if the next commit (https://github.com/ThrowTheSwitch/Ceedling/pull/739/commits/e9f35da1497df03fcf25a5fcac66954db169fb44) fixes this, but it does not look like it.

swaldhoer commented:
Ah that's weired, there is a pre-release for
- commit https://github.com/ThrowTheSwitch/Ceedling/pull/739/commits/ba45d2c0d2ba7da67ae09eb3b1d5e32f2561ded1 :: https://github.com/ThrowTheSwitch/Ceedling/releases/tag/1.0.0-ba45d2c
but not for 
-commit https://github.com/ThrowTheSwitch/Ceedling/pull/739/commits/e9f35da1497df03fcf25a5fcac66954db169fb44 :: does not exist at https://github.com/ThrowTheSwitch/Ceedling/releases (and it is not available via https://github.com/ThrowTheSwitch/Ceedling/tags, but there is no pre-release gem for it).

So in short: I already tested the latest version, and it does not work.

mkarlesky commented:
@swaldhoer There were two prereleases that added detailed versioning. The first did not include vendor support; the most recent prerelease does. I tested it a bunch, and it was working. Can you share more details of what prerelease you're using and how you're installing a prerelease?

The key detail is a GIT_COMMIT_SHA file containing the short SHA string in the root of the Ceedling application that's created when the gem is built in Github Actions. That is, this only works by installing the gem. If that file does not exist, no extra version information is displayed. You can verify if that file exists in your gem. You can also trick Ceedling to produce version information by causing that file to exist in the root of the application.

(Incidentally, not every commit necessarily creates a prerelease. Or, rather, I sometimes cancel a build if only documentation or other non-code changes are merged.)

swaldhoer commented:
I used the latest pre-release [1.0.0-ba45d2c.](https://github.com/ThrowTheSwitch/Ceedling/releases/tag/1.0.0-ba45d2c).
I basically added it via `ceedling new --local`.

I think the problem is then just, that `GIT_COMMIT_SHA` file does not get copied into the vendoring dirctory?

mkarlesky commented:
@swaldhoer The sha file does get copied for a vendored version of Ceedling (but maybe there's a bug in how it's found in the vendoring scenario). You should be able to run `gem info ceedling` to get the installation path. Please look in there and see if the GIT_COMMIT_SHA file is present.

Letme commented:
I pulled latest release fcf7c82 and installed it with `gem install ceedling.gem` and the `ceedling --version` outputs version with sha in existing project:

```
$ ceedling --version
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
🌱 Welcome to Ceedling!

   Ceedling => 1.0.0-fcf7c82
      CMock => 2.5.4
      Unity => 2.6.0
 CException => 1.3.4
```

Do you have maybe a submodules version of the ceedling and you default to that?

mkarlesky commented:
@swaldhoer @Letme I'm going to quick add a flag to `version` that prints _which_ Ceedling (its path) is running. All of that is well managed internally already. It's easily accessible. It seems this will be a helpful troubleshooting tool generally.

swaldhoer commented:
@mkarlesky it's working as expected, I made a mistake while vendoring - all good!

Letme commented:
Its been some time since I tested pre-release, but now I have a problem that all defines are stripped from the mocks. I do not think that was happening before. Did something change?

mkarlesky commented:
@swaldhoer Your experience with version complications inspired some changes to `ceedling version` and build logging to help make it clear what version of Ceedling is running and from where (available in the latest prerelease). It can get a bit complicated or misleading. Hopefully these changes will help prevent confusion in the future and make troubleshooting much easier for all involved.

swaldhoer commented:
Glad to hear that, I will have a look at the latest pre-release tomorrow and test it!

Letme commented:
Ok, my issue was indeed that `:all` in test_preprocessor, which I now changed to `:none`. Which is strange, because I think before it was only true/false and I probably had true, so that I changed it to all, but because setting defaulted to `:none` that was the reason it worked. Meanwhile configuration checking improved in Ceedling, so I got tracked to wrong path.

Test it with also latest release.
